include/boost/hana/sort.hpp

```diff
@@ -68,82 +68,171 @@ BOOST_HANA_NAMESPACE_BEGIN
+            std::size_t r1,
+            std::size_t ...r>
+        struct merge<
+            Pred,
+            true,
+            std::index_sequence<l0, l...>,
+            std::index_sequence<r0, r1, r...>
+        > {
+            using type = typename concat<
+                std::index_sequence<r0>,
+                typename merge<
+                    Pred,
+                    (bool)Pred::template apply<r1, l0>::value,
+                    std::index_sequence<l0, l...>,
+                    std::index_sequence<r1, r...>
+                >::type
+            >::type;
         };
 
+        template <
+            typename Pred,
+            std::size_t l0,
+            std::size_t ...l,
+            std::size_t r0>
+        struct merge<
+            Pred,
+            true,
+            std::index_sequence<l0, l...>,
+            std::index_sequence<r0>
+        > {
+            using type = std::index_sequence<r0, l0, l...>;
+        };
 
-        template <typename Pred, typename Result, std::size_t ...T>
-        struct insertion_sort_impl;
+        template <typename Pred, typename Left, typename Right>
+        struct merge_helper;
 
-        template <typename Pred,
-                  std::size_t Result1, std::size_t ...Result,
-                  std::size_t T, std::size_t ...Ts>
-        struct insertion_sort_impl<Pred, std::index_sequence<Result1, Result...>, T, Ts...> {
-            using type = typename insertion_sort_impl<
+        template <
+            typename Pred,
+            std::size_t l0,
+            std::size_t ...l,
+            std::size_t r0,
+            std::size_t ...r>
+        struct merge_helper<
+            Pred,
+            std::index_sequence<l0, l...>,
+            std::index_sequence<r0, r...>
+        > {
+            using type = typename merge<
                 Pred,
-                typename insert<
-                    Pred, T, (bool)Pred::template apply<T, Result1>::value,
-                    std::index_sequence<>,
-                    Result1, Result...
-                >::type,
-                Ts...
+                (bool)Pred::template apply<r0, l0>::value,
+                std::index_sequence<l0, l...>,
+                std::index_sequence<r0, r...>
             >::type;
         };
 
-        template <typename Pred, std::size_t T, std::size_t ...Ts>
-        struct insertion_sort_impl<Pred, std::index_sequence<>, T, Ts...> {
-            using type = typename insertion_sort_impl<
-                Pred, std::index_sequence<T>, Ts...
-            >::type;
+        template <std::size_t Nr, typename Left, typename Right, typename=void>
+        struct split;
+
+        template <
+            std::size_t Nr,
+            std::size_t ...l,
+            std::size_t ...r,
+            std::size_t r0>
+        struct split<
+            Nr,
+            std::index_sequence<l...>,
+            std::index_sequence<r0 ,r...>,
+            typename std::enable_if<Nr!=0>::type
```

Comment:
You need the `enable_if` because otherwise, when `Nr == 0` AND the second `std::index_sequence` is empty, you get an ambiguity. Is that right?

Reply:
The problem is that these two specializations are not strictly ordered because:
- Template 1 cannot match N==0 AND empty second sequence, matched by template 2
- Template 2 cannot match N!=0, matched by template 1

Adding enable_if makes sure they will never be both valid candidates in an overload resolution.

As you stated, an ambiguity would occur if N==0 and the second sequence is *not* empty (I'm assuming there is a typo in your comment).

Reply:
Feel free to add this in the comments, so this information is not lost...

Reply:
Done